### ISSUE TITLE
Moving Qrimatic into Qri core

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#1787]
## Questions

implementation questions:
- What mechanism should we use for creating run ids? `run` package uses uuid, `workflow` package uses `go-cid` `cid.NewCidV1`
- We have lib methods defined for deploying, updating, and removing a workflow, but nothing specified for if you just want to "undeploy" or "unschedule" the workflow. Unless this is supposed to be folded into another method and specified using params (which we have been trending away see `ConnectToPeer` and `DisconnectFromPeer`), should we add `UndeployWorkflow` to our list of automation methods? A user might have a case where they want to pause the workflow, but still keep all of the workflow information, but they may also have a case where they no longer want a particular workflow to show up in any lists/remove it completely. We would need a mechanism for pausing and a mechanism for removal.
- Does `workflow.EndTime` include the time it takes to execute the hook? Or is it synonymous w/ `run.EndTime`? I'm leaning toward the latter. Same question, but is `Duration` the time it takes to run the whole workflow, or just execute the run. Again leaning toward the latter, we just need to make it clear that's what it represents.

conceptual questions:
- We need a run log store, correct? Don't fully understand if the log of a run is different than what exists in our `run.Run` structure, or if it contains more information than what would exist in a `Run`. EG, if a `Hook` fails, that seems like it should exist in a log of the run, but that knowledge exists outside of a run, it exists in the workflow. So maybe I'm just using the word `run` incorrectly, or we are overloading it in the context of qri. Workflow output?  Regardless, we need a cache for this. Are we setting up a new store, or does this belong as a part of a different system.

Needs more research before I have a real question:
- we should move `RunInfo` into `run` package, however some thing that exist on `RunInfo` don't exist on a `run`. Need to do more research to understand what exactly belongs on a`RunInfo`
     - `run.State` in run has `Message`, `Status`, `Duration`, `Steps`
     - `workflow.RunInfo` has `WorkflowID`, `Error`, `LogFilePath`
     - both have `ID`, `Number`, `Start`, `Stop`
     - I think we can get rid of `RunInfo`, and just use run, need to see how it's worked
- Should `hooks`, `triggers` have their own packages, ala `run`? Then `workflow`'s responsibility is just defining & storing a workflow (which is made up of hooks, trigers, and runs).
- basically seeing that we have all these packages but only one place where we store shit (Aside from transforms technically being saved in the dataset) 
- this should come as no surprise, but `Hook` feels the least thought out to me. There is an interface, a struct, a list of types. If there is an interface, why do we have a list of types. Shouldn't we instead have various structs that satisfy the hook interface, and use the hook methods to interact with it? This would require adding more methods to the interface, such as a method that actually runs the hook. Or have a generic `HookInfo` that other hooks can extend.
- perhaps larger `automator`, keeps track of stores all the possible triggers, runs, and hook executors. `scheduler` (trigger executor?), `runner`, and `hookexecutor` are subsystems that are "muxed", each having their own subsystems that maybe run depending on the kind of trigger, run, or hook, specified. EG, one of the subsystems in the `trigger executor` would be `cron`, another may be `webhook`, another may be `downstream dep`. Each with its own loop listening for its own triggers (with the `downstream dep` trigger listening to dataset update events, perhaps). When triggered, it either sends info down a channel, or we use an event bus, to trigger a run from the `runner` subsystem. When that completes, it either sends info down a channel, or we use the event bus, to trigger a hook. Each subsystem should use the workflow store to understand what types of actions it needs to do. AKA, the trigger is not responsible for letting the runner know if it should run a `starlark` transform or (in the future) some other kind of transform. The runner should look up the workflow and figure out what it needs to do. This may not be the best move, but it then divides the work in this way: the system relies on events for timing and the store for information. Main questions... what does the store rely on to update? Direct access? Does it also rely on events. EG, we have a field `Number` on the workflow struct. Is the workflow store listening to events as well, and when a trigger event comes across the wire, does it update itself. Or does the runner subsystem "manually" adjust the `Number` and `Start` time, and does the 
- scheduler - Is the scheduler package only responsible for cron triggers? Should there be a separate package for other kinds of triggers? A looping of a different subsystems that wait for webhooks, for example, or for tracking when another dataset updates? Thinking they should be the same thing? In which case, `scheduler` is a misnomer. Should be... runner? tracker? Could also be 3 totally diffeent systems that rely on events. `tracker` (pays attention to triggers, notifies the next system when a trigger has fired), `runner` (runs the actual transform, emitting events as things progress, succeed, and fail), `completioner` (name... pending, listens for specific events and fires off appropriate hooks)
- cron is added in NewInstance, workflow store & scheduler. Perhaps should be multiplexed the way qfs is. scheduler mux, one of which is cron. Or, if scheduler w/ config needed and the scheduler contains different ways to create. Intead of `NewCronScheduler`, we have `NewScheduler`, which for now only adds a cron scheduler. & should inst really be where the store lives, or should that only be accessable via the scheudler. I need to think through. TBH probably instance is fine? are their workflows without a scheduler? I don't htink so. 
- debate: should triggers, workflows, and hooks be stored separately by workflow id then? Gut instinct is no, one store all systems can coordinate on

- trigger monitoring will be the most robust and varied. Needs to keep track of many different possiblities, different kinds of input, probably a big overhead when starting up the system, between establishing a list of what needs to be cared about, but also immediately fire off updates that have been delayed because the system has ben closed down
- runs are the most uniform, once the trigger alerts that a run must happen, the way the runs all happen is the same. Until we introduce different languages or ways to run a transform, this will probably be the most stable. Bulk of the work isn't even happening here, really, it's just keeping track of the number, length, and status, and storing those. Possible job is to store the Run log itself, besides also update the workflow info, could be that this is also where we gather up the information needed for the hook & just send over the information to the hook subsystem to figure out on its own.
- hook system could possibly be v dumb, just take the information given, and send off, no loading, no notations... except if the hook fails, we would probably want to have that in a run log, no? 

- what is the purpose of a workflow? What information is it storing, should we be storing it. Should we instead be recording information about it in the logbook, and be getting the workflow information as a result of querying the logbook.
- if workflow is made up of:
  - dataset info
  - trigger info
  - run info
  - hook info
  - additional metadata... like number of runs

## Package purposes:

- `run` 
	- `State struct`: define a cacheable data structure that describes the execution of a transform
	- `Status`: enumerates al possible execution states of a transform script or step within a script
	- `StepState struct`: describes the execution of a transform step 
	- ? `Info struct`: a lighter weight version of a run, used primarily for listing. Contains much a the same information as a run, except the `Steps`
- `scheduler`
	- `Scheduler` - interface for a cron scheduler
	- `RunWorkflowFunc` - function for executing a workflow. Scheduler takes care of scheduling workflow execution, and delegates the work of executing a workflow to  RunWorkflowFunc
	- `Cron struct` - coordinates the scheudling of running workflows at specified periodicities with a provided workflow runner function
- `transform` - 
	- `tranformer` - holds dependencies needed for applying a transform, methods include `Apply`, which applies a transform script to a target dataset
	- `startf` - implements dataset transformations using the starlark programing dialect 
- `workflow`
	- `Workflow` - defines a cachable structure used to describe the state of a workflow. A workflow can be made up of a list of `Triggers`, `Transform.Steps`, `Hooks`, and `run.State`. A workflow should have all the information needed to schedule, run, and complete a transform.
	- `Store` - defines a structure used to store workflows
	- `events.go` - defines events published in the scheduler.
	- `Hooks` - interface for a type of hook. A hook is a user-defined action that should be run, conditionality, when a tranform complete
	- `HookInfo struct` - contains the information needed to run a hook after a workflow has run 
	- `Triggers` - interface for a type of trigger. When a trigger fires, a workflow runs.
	- `TriggerInfo struct` - contains the information needed to keep track of a trigger.   

### lib methods
- `Apply` - apply a transform script to a dataset, i'm not sure apply should exist here. should probably be on dataset. only argument for is coordinating if the workflow is already running.
- `ListWorkflows` - Get a list of workflows
- `GetWorkflow` - get a specific workflow
- `DeployWorkflow` - schedule a workflow to run
- ? `UndeployWorkflow` - unschedule the workflow, but keep it around
- `SaveWorkflow` - update a workflow
- `RemoveWorkflow` - remove a workflow (delete it fully)

#### integration test flow:
- Deploy workflow a
- Deploy workflow b
- get workflow a
- list workflows
- save worklow a
- get workflow a
- undeploy
- list
- remove
- list

**Comment by b5** [comment]
> My biggest confusion is around run logs, what they are supposed to track, and how we want to store them.

This question is about run log _retention_. When the run log subsystem is finished, it will act as a _sink_ for `events` collated by a `runID`. Each `event` is an event bus event published by the `transform` package while a transform is being applied. persisting run logs allows us to "play back" a transform to see the output & timing of each steps.  Without run log persistence, we can't show users how their scripts failed unless they are watching them happen in real time via the UI.

The events emitted by the transform package have  been designed to serialize nicely to newline-delimited JSON, so a local implementation might be set up as a file-per-runID of `NDJSON`, plus a file for the index.

One challenge with run logs: they can get big. Recording lots of output can take up space, so we want to define the run log subsystem with a configurable max-size, default _retention strategy_, and API for manually deleting log files to free up space. #1748 describes the `runLogMaxCacheSize` configuration value that would govern how many logs to retain in MB.

**Comment by ramfox** [comment]
>  Run is strictly about transform execution, while workflow holds the broader execution context.

Yea this was my initial reaction. But when I thought about it in practice, I didn't really understand what that meant. We don't start a workflow until it is triggered, so the start of the run is the same as the start of the workflow. The only difference is the end time with the hook. But what are we responsible for tracking on the hook side? If we are sending off an email notification, is it the time the notification gets sent? The time the server receives the email? Same with sending off a webhook, is it the time we trigger the webhook or the completion of the sending of the webhook. The only exceptions I can think of is if there is some kind of validator or extra processing that happens on completion, in which case I will fully buy that the full workflow duration is useful.

**Comment by ramfox** [comment]
Workflow is made up of fields that are changed at different rates
1) permanent, changing one of these fields changes the workflow fundamentally
 - ID
 - DatasetID
 - Created
2) Rarely changed - changing one of these fields would mean a large change in permissions
 - OwnerID (assuming this can be "added to" in the future, where a workflow can be owned by multiple people who have permission to edit it
3) User changed - these fields can be changed without changing the identity of the workflow
- Options
- Enabled
- Triggers
- Type
- OnComplete
- Type
4) Changed every run - automated changes that happen ever run
- RunCount
- LatestStart
- LatestEnd
- CurrentRun
- VersionInfo
5) changed multiple times per run - automated
- Status

Trying to mentally separate what belongs in `workflow` and what belongs in the `scheduler`. State belongs in the `workflow.Store` and how we react to that state belongs in the `scheduler`. So what is it that we need when we "deploy" "undeploy" "update" "remove" etc. Want to move much of this work into the `Store`, so the store will create a new workflow when we "create" one, while "deploy" just ensures that the workflow is enabled. 

If the scheduler is active, it should listen to the events of the store, see that a "deploy" event has passed, and react to it accordingly.

**Comment by ramfox** [comment]
Question. What does it mean to deploy a workflow, when you don't have a daemon running. I'm imagining a scenario where a user "deploys" a workflow from the cli, if this is something we are allowing, or if we have a daemon running, but the scheduler isn't listening for triggers (again if this is something we are allowing).

3 potential things happen when we "deploy"
- create a new workflow
- "enable" the workflow (workflow isn't paused)
- run the workflow immediately

Each of these may not be what a user wants. The first item is easier to understand, if there is already a workflow with that id, then we don't need to create a new one. The second is implied by "deploy", if you are "deploying" the workflow you are enabling it. The last is a question. So, does "deploy" mean run now in all cases? Is this a flag that we need to pay attention to. 

We should probably isolate these three things, and then combine back in convenient ways for the user.

At the workflow level, workflow stores should be able to do:
- Create
- Update
- Get
- Remove
- Deploy
- Undeploy

We already have a `RunWorkflow` func on the scheduler.

at the lib level, we should have options for "DeployWorkflow" that allows a user to run now. "DeployWorkflow" tries to `workflow.Get`, if nothing exists, `workflow.Create`, then `workflow.Deploy`, to enable the dataset, if the `DeployParams` have `runnow==true`, run `scheduler.RunWorkflow()`... This may come into conflict with cron triggers, as the scheduler will auto run that regardless, as soon as the workflow is enabled.

However, once the scheduler has been launched, any workflows that are outstanding will run when we create the instance to run this workflow. How do we prevent that? This will be a problem for every command, unless we prevent the mounting of the scheduler unless explicitly called. And if they are separated, can you run the scheduler and the normal qri server at the same time?

If we don't have an expectation of "deploy" == run immediately via the cli (unless the scheduler is already running), than this isn't a problem.

**Comment by b5** [comment]
#### Run length vs. Workflow length

> Yea this was my initial reaction. But when I thought about it in practice, I didn't really understand how they were different. We don't start a workflow until it is triggered, so the start of the run is the same as the start of the workflow. The only difference is the end of a run vs the end of a hook.

I think you're right: runs should describe the execution of an entire workflow when needed, matching the duration of an entire workflow start/stop, and capturing logs from completion hooks if they emit events.

I say "when needed" because this doesn't apply to dry-runs, where completion hooks are intentionally left unexecuted. We'll need to attach a "run type" field of some kind to the `RunStart` event to be able to distinguish between dry runs and normal runs that cover the entire transform execution.

This is the better approach for at least two reasons: the first is it's just easier to think about once you get your head around it: they're _workflow_ runs, **not** transform runs. The second is the added debugging capacity that comes from recording events emitted by completion hooks It'll be _way_ easier to debug failed completion hook execution if we just capture the output in the same run log.

**Comment by b5** [comment]
> What does it mean to deploy a workflow, when you don't have a daemon running?

We should print a warning that a daemon needs to be running to keep automating the workflow, and not do anything until `qri connect` executes & `automationEnabled: true` is set in configuration. This is the behavior I've currently been working with, and TBH it's not so bad locally. It means that whenever I run `qri connect` I just kinda sit & watch as the scheduler knocks out all the automations I've missed. It does feel a little like Christmas.

That said, it does have a frustrating downside: if an automation is scheduled to run every hour, and you only run `qri connect` once a day, you're only getting one update, and transparently skipping 23 scheduled runs. I'm fine with this,  and think we just need to warn users. qri.cloud will exist for people who want full autopilot, and we'll write docs on how to deploy qri to a server for those who'd like something in the middle

**Comment by ramfox** [comment]
Start of `automation` spec

```
new package tree
├── run
├── trigger
├── completion
├── transform
└── workflow
    ├── events.go
    ├── hook.go
    ├── run.go
    ├── store.go
    ├── store_file.go
    ├── store_file_test.go
    ├── store_test.go
    ├── trigger.go
    ├── workflow.go
    └── workflow_test.go
```

The automation package is responsible for defining the `orchestrator` system. `orchestrator` is responsible for orchestrating the automation process.

```
package automation

type Applier interface {
 	// Apply runs a transform on a dataset to create a new version of the dataset
	Apply(ctx context.Context, target *dataset.Dataset, runID string, wait bool, scriptOut io.Writer, secrets map[string]string) error
}
type Orchestrator struct { /* ... */ }

// NewOrchestrator is responsible for constructing an Orchestrator. The Orchestrator creates a `workflow.Store`, subscribes to `event.ETTriggerAutomation` events, and sets up any `Completion` services. The Orchestrator relies on an `Applier` function that it calls to actually run the expected automation
func NewOrchestrator(cfg config.Automation, bus event.Bus, runTransformFunc Applier) (*Orchestrator, error)

// StartTriggers constructs and starts subsystems that listens for various triggers and emit `event.ETTriggerAutomation` events
func (o *Orchestrator) StartTriggers(ctx context.Context)

// StopTriggers tears down any systems that were created in `StartTriggers`
func (o *Orchestrator) StopTriggers()

// handleTrigger is the handler called when an `event.ETTriggerAutomation` is fired. It is responsible for creating a RunID, Advancing the workflow, calling the Applier, emitting any necessary `completion` events, cleaning up the workflow, and determining the exit code.
// workflows should run sequentially, so `handleTrigger` should block
func (o *Orchestrator) handleTrigger()

// RunWorkflow is responsible for manually triggering a workflow run, by emitting an `event.ETTriggerAutomation` event
func (o *Orchestrator) RunWorkflow(ctx context.Context, wf *workflow.Workflow) error

// DryRun runs only the `Apply` part of the workflow, not creating a new version of a dataset, nor sending off any completions
func (o *Orchestrator) DryRun(ctx context.Context, ds *dataset.Dataset) error
```

```
package workflow

// Workflow holds both the instructions on how to run an automation, as well as the state of the latest/current run
type Workflow struct { /* ... */ }

// Advance sets the workflow up for a new run
func (w *Workflow) Advance() error

// Complete records any information needed to note at the end of the run
func (w *Workflow) Complete() error

// Store handles the persistence and listing of Workflows. Store implementations must be safe for concurrent use
type Store interface {
	Updater
	Lister
}

// Updater handles the persistence of Workflows
type Updater interface {
	// Creates a new undeployed workflow
	Create(wf *Workflow) (*Workflow, error)
	// Updates a workflow 
	Update(wf *Workflow) error
	// Remove removes a workflow from the store
	Remove(id string) error
	// Get a workflow from a workflow id
	Get(id string) (*Workflow, error)
	// Deploy enables the workflow to be run by an orchestrator
	Deploy(id string) error
	// Undeploy disables the workflow from being run by an orchestrator
	Undeploy(id string) error
}

// Lister lists workflows
type Lister interface {
	// List lists all the workflows in the store
	List(ctx context.Context, offset, limit int) ([]*Workflow, error)
	// ListByStatus lists all the workflows of a particular status
	ListByStatus(ctx context.Context, status Status, offset, limit int) ([]*Workflow, error)
}
```

```
package trigger

// Watcher watches for triggers and emits an `event.ETTriggerAutomation` event
type Watcher interface {
	Start(ctx context.Context) error
	Stop() error
}

// NewWatcher creates a watcher that watches for a variety of triggers
func NewWatcher() (*Watcher, error) {
	// creates systems for watching for triggers, eg cron watcher. doesn't start watching
	return &TriggerSystem{}
}

// TriggerSystem needs better name, responsible for managing the different trigger subsystems
type TriggerSystem struct { /* ... */ }

func (t *TriggerSystem) Start(ctx context.Context) error {
	t.cron.Start(ctx)
	t.dependencies.Start(ctx)
	t.webhook.Start(ctx)
	// or
	// for _, watcher := range t.watchers {
	// 	watcher.Start()
	// }
}

// Stop shuts down any subsystems that are listending
func (t *TriggerSystem) Stop() error {
	t.cron.Stop()	
	t.dependencies.Stop()
	t.webhook.Stop()
	// or
	// for _, watcher := range t.watchers {
	// 	watcher.Stop()
	// }
	// wait until all closed, then return
}

// Cron coordinates the scheduling of running workflows at a specified perodicity, emitting an `event.ETTriggerAutomation` event when at the specified perodicity
// it queries the workflow.Store
type Cron struct { /* ... */ }

// NewCron creates a Cron with a check interval
func NewCron(workflows workflow.Lister, pub event.Publisher, checkInterval time.Duration) Watcher

func (c *Cron) Start (ctx context.Context) error {
	// at the given checkInterval, query the workflow.Lister, getting a list of workflow, if any have cron triggers check if they should be run
	// if so, emit the `event.ETTriggerAutomation` event w/ the workflow id
}

// disrupt the cron check loop
func (c *Cron) Stop() error

type Trigger interface {
	Advance() error
	// serializes Trigger for storing
	ToMap() map[string]string
}

func TriggerFromMap(m map[string]string) *Trigger {
	// switch on m["type"] to deserialize trigger
}

// CronTrigger represents a specific kind of trigger that repeats at a set periodicity
type CronTrigger struct { /* ... */ }

func (ct *CronTrigger) Advance() error
func (ct *CronTrigger) ToMap() map[string]string
 ```

**Comment by Arqu** [comment]
Nice, breakdown for a lot of things. I've read from the start to get fully caught up, was a great crash course in the current state of things with qrimatic and its nuances.

Anyways regarding the last bit, this should fly well with the cloud side too as the interfaces are neatly aligned with the data-store pattern we use in qri/cloud.

**Comment by ramfox** [comment]
That's great to hear! Thank you for taking the time to read through.

**Comment by ramfox** [comment]
Major question: are we in a place to rely on dataset's InitIDs?

**Comment by ramfox** [comment]
Okay decided to split this into a few parts:
1) WorkflowStore & Orchestrator
- have lib create an Orchestrator whose job, for now, is only to create a workflow store & be the proxy between the instance and the workflow store, eg the lib method would call `scope.Orchestrator().RemoveWorkflow()` to remove a workflow, rather than having access to the workflow itself.
- Workflow creation/removal/listing/retrieval can be tested separately from any automation
(starting here will also help me feel like I'm making tangible progress as we iterate on portions of the spec that are less fleshed out)

2) Add triggers to orchestration, and add starting orchestration to the connect command
- need to add a `config.Automation` section, w/ a field to enable/disable automation
- need to figure out if we are using the config to allow users to set what kinds of triggers they want to listen for, or if we have a standard set of triggers we always listen for always. Probably some combo, we will internally use some `NewAutomationFromConfig` that takes a config & doesn't take a list of pre-made triggers, and `NewAutomation` that takes a list of TriggerListeners.
- Add `orchestrator.Start()` `orchestrator.Stop()` that starts the trigger listeners and tears down the trigger listeners, respectfully
- adjust lib & `qri connect` accordingly

3) ensure runs still work and emit events as expected in new orchestrator system

4) completers 
- list of events that orchestrator should emit based on expected completers
- system that listens to & handles complete events
- tbh have this a bit more fleshed out than even triggers, because we are truly making completers opaque to the orchestrator, which means they need to be created and passed in, so there is a good chance they won't exist in any configuration. However, still need to talk this out.

**Comment by ramfox** [comment]
When a user calls `Save` w/ a transform, is that considered a run? Do we track that as the last run? Is that added to the run count?

What's the relationship between a workflow and the actual transform script. How does a user make a draft transform?

How to associate a Run w/ a workflow. Since a run can also exist without a workflow, do we instead associate it by dataset id?

**Comment by ramfox** [comment]
Currently planning projects in this document:
https://hackmd.io/4H9ZFE9OQ7KY3ytteZFmhw?edit

**Comment by b5** [comment]
> When a user calls Save w/ a transform, is that considered a run?

Yes, but I don't think the run store should persist it. We talked a bit in our last design session about how there are different "kinds" of runs. After our last design session I think we tacitly arrived at the conclusion that any run we save to the run store is an _artifact of executing a workflow_, meaning runs must be owned by a workflow. But on the other hand when we're running "save" or "dry run" we still need to collect an event stream collated by a `runID`. Frontend needs to get these over the websocket for stuff like dry runs, CLI needs to know which events to pipe to `stout` for both `apply` and `save`. 

I think the distinction comes in what runs we persist, and in order to persist a run in a `run.Store` it needs a `workflowID`.


> What's the relationship between a workflow and the actual transform script. 

Workflows are defined on dataset identifiers, and the dataset owns the transform component. Workflows shouldn't retain any details about transforms. So in a scenario where a user creates a dataset with `qri save` that has a transform, then creates a auto-updating workflow on that dataset, but then runs `qri save --remove tf`, the next time the workflow runs it should exit with `ErrNoTransformComponent`.

> Is the process of making a "Draft transform" only something that exists on the frontend.

**Comment by ramfox** [comment]
closed by 
#1875 
#1853
#1862
#1856 
#1855 
#1854 
#1853 
#1852 
#1851 
#1849 
#1848 
#1847 
#1845 
#1841 
#1838 
#1835 
#1834 
#1822
#1814 
#1803
#1797 
#1795
